### ISSUE TITLE
HeatMap selecting particular point

### ISSUE THREAD

**Issue by kalikadze** [swharden/ScottPlot#645]
Hello guys, 

I would like to ask you if is possible somehow select particular point on HeatMap. Better, double click on point to get X, Y coords of this point.

Additional question. Is possible to have, let's say circle heatmap where points outside of circle will not affect with their values the intensity of of color points inside circle. In other words, I don't want points from outside the circle to have impact on the coloring inside the circle. Is possible to have them for example vith value Double.NaN?

thank you very much.
BR.
kali

**Comment by swharden** [comment]
Hey @kalikadze, I think I can help you out here!

## Determine Mouse Position When Clicked

Catch the `MouseClicked` event triggered by the FormsPlot control. You could do the same thing with the double click event.

![image](https://user-images.githubusercontent.com/4165489/102702806-f02ac700-4234-11eb-9297-78d7b887bd84.png)

```cs
public Form1()
{
    InitializeComponent();
    formsPlot1.plt.PlotSignal(ScottPlot.DataGen.Sin(51));
    formsPlot1.plt.PlotSignal(ScottPlot.DataGen.Cos(51));
    formsPlot1.Render();
}

private void formsPlot1_MouseClicked(object sender, MouseEventArgs e)
{
    double x = formsPlot1.plt.CoordinateFromPixelX(e.X);
    double y = formsPlot1.plt.CoordinateFromPixelY(e.Y);
    Text = $"Clicked X={x} Y={y}";
}
```

![coords](https://user-images.githubusercontent.com/4165489/102702837-54e62180-4235-11eb-9d03-ef653b7bfcd3.gif)

The window title shows the mouse position after every click

## "Hidden" Values in Heatmaps

It sounds like your goal is to plot a rectangular heatmap where not every square has data. You could get creative with a custom colormap (where the lowest color is `Color.Transparent`), but that sounds somewhat complex. 

I like the idea of `Double.NaN` being ignored and not rendered though. @Benny121221 you work a lot with these heatmaps, what do you think? I see #667 is being actively developed, maybe this is a good feature to sneak in with that PR?

---

I'll close this issue but @Benny121221 feel free to open a new issue / discussion / PR if you want to discuss the heatmap topic further

**Comment by Benny121221** [comment]
> I like the idea of Double.NaN being ignored and not rendered though.

I'm unsure what you mean, do you want NaN values to be mapped to the lowest intensity colour? Or mapped to transparent? Currently, the occurrence of any NaN value breaks the heatmap (every pixel is mapped to the lowest intensity colour). Personally 

I'm unsure that either makes sense for NaN, and I think the best idea is to replace them with a more appropriate value (the minimum of the heatmap if one wants the lowest intensity colour, and a sentinel such as `double.NegativeInfinity` if one wants transparency). Note that if one wants transparency one needs to use the `transparencyThreshold` parameter. You could also set every `NaN` value to `double.NegativeInfinity` and use `scaleMin` if you want the clipping behaviour.

I'm very hesitant to hand down an interpretation of NaN as a real (or in the case of infinities, a hyperreal) number, especially if that interpretation would conflict with the rest of the library (where NaN is treated as an error). For example, I believe NaN came up in a previous issue and it broke someones signal plot. The solution there was not to quietly treat NaN as just another number (or even remove the NaN value and continue plotting as normal), it was to add diagnostic capabilities to prevent NaN from finding it's way into a signal plot #546.

In case you're curious, the reason NaN breaks heatmaps is because by definition (in IEEE 754) `NaN > x` and `NaN < x` will always return false, for all values of x. As you can imagine that makes finding the maximum and minimum difficult, which means the heatmap cannot be normalized properly.

**Comment by swharden** [comment]
>> I like the idea of Double.NaN being ignored and not rendered though.

> I'm unsure what you mean, do you want NaN values to be mapped to the lowest intensity colour? Or mapped to transparent?

I was thinking normalized intensities would contain NaNs for empty data points, so a NaN check before `DrawRectangle()` would allow only non-NaN points to be rendered

> set every NaN value to double.NegativeInfinity and use scaleMin if you want the clipping behaviour

That sounds like a fine solution as well!

> I believe NaN came up in a previous issue and it broke someone's signal plot.
> The solution there was...

How I treat NaN in data has changed a bit with time and experience. The direction I'm going now is that I try to throw an exception ASAP if NaN is detected in plottables (like Signal and Scatter) which do not support them.

I think it would be ok to disallow NaN in some plottables but permit others, so it's up to you if you prefer NaN, Infinity, or something else (`bool[,] Visible`?)

> by definition (in IEEE 754) NaN > x and NaN < x will always return false, for all values of x. As you can imagine that makes finding the maximum and minimum difficult, which means the heatmap cannot be normalized properly.

I write safer (albeit marginally slower) code using `Math.Min()` and `Math.Max()` for comparisons where input data could contain NaN. Those methods return NaN if either of the inputs is NaN, so NaNs propagate through the system. If you decide that NaN is forbidden, this is a good way of simultaneously finding min/max and detecting if `NaN` is in an array.

> the best idea is to replace [NaN] with a more appropriate value [Infinity]

I like this idea! This way we _accept_ NaN but replace it quickly and work with NegativeInfinity. What do you think?

**Comment by Benny121221** [comment]
> > the best idea is to replace [NaN] with a more appropriate value [Infinity]
> 
> I like this idea! This way we accept NaN but replace it quickly and work with NegativeInfinity. What do you think?

I meant replace NaN in user code... That's why I listed two possible "appropriate values", negative infinity, and the min of the given data. I think depending on the desired behaviour, both could be appropriate.

Worth noting that negative infinity would break heatmaps as well without a user-defined `scaleMin`, so negative infinity doesn't solve the problem where heatmaps cannot determine a real min and max, which leads to them rendering as all one colour (in the case of NaNs and positive infinity it's the minimum intensity, in the case of negative infinity its the maximum intensity, and in the case of negative and positive infinity I'm not actually sure as I didn't bother to test which effect is "stronger").

So in my opinion, if NaN is silently replaced with negative infinity it still leads to a broken plot in most cases (those without a non-null `scaleMin`), while still having the problem where we axiomatically define something which is by definition not a number to be a number.

Negative infinity _does_ have the benefit that it's still a "wonky" number, and while debugging a user who sees negative infinity is going to think "huh, I guess I gotta do some checks on what numbers I pass in", whereas if they see the minimum of the numbers they passed in they're unlikely to notice it.

I still think the best solution is to not accept NaN, throwing immediately if possible (and if not add a `ValidateData` method to scan for NaN when called). Heatmaps already have support for transparency so I don't see the need to create a (in my opinion) hacky workaround. That being said, I'm noticing that the transparency threshold acts as if it's much lower than it is, I made a PR to fix it: #668

**Comment by swharden** [comment]
Hey @Benny121221, thanks for your input! It's a recurring pattern that missing data and magic floating-point numbers really complicate plotting...

### What do to with NaN

> if NaN is silently replaced with negative infinity it still leads to a broken plot

> I still think the best solution is to not accept NaN, throwing immediately if possible

I agree! I'm guessing the best thing to do is check every value of `intensities` for NaN right at the top of this method? Depending on what you think you can add it in #668 or I can add the check myself after merging it.
https://github.com/swharden/ScottPlot/blob/5cbf9262001a539c34bb9a5b9aca691b5ef483a7/src/ScottPlot/Plottable/Heatmap.cs#L37-L38

### The OP's goal

> Heatmaps already have support for transparency so I don't see the need to create a (in my opinion) hacky workaround

@kalikadze originally asked "let's say circle heatmap where points outside of circle will not affect with their values the intensity of of color points inside circle". What is the recommended way to display such this data into the heatmap? Create a 2D array of values and set all the "empty" values to a number lower than `TransparencyThreshold`?

Until I hear otherwise I'll assume @kalikadze is monitoring control rod temperatures in a nuclear reactor. It's the only reason I can think of to have a circular heatmap 🙃

**Comment by Benny121221** [comment]
> Create a 2D array of values and set all the "empty" values to a number lower than TransparencyThreshold?

Yep, that's the best way to do it. Please note that `TransparencyThreshold` is broken, this is fixed in #668.

**Comment by kalikadze** [comment]
Hello guys, 

thank you for your perfect inputs! Meanwhile i started to finding solution for my problem, and i finish up with something similair like you proposed there :) 

I tried to find some way how to vizualize data from silicon wafers. So, i don't messing up  with a nuclear reactor :D 
![image](https://user-images.githubusercontent.com/10723981/102722849-38fd8100-4304-11eb-8135-9ba8c2136408.png)
I did it as you said, yesterday evening :) 

Started with a double  array [,], init with values double.MinVal (using this like null or invalid data), then I filled up with my measured data (circular shape), All invalid values are set sligthly below min value (depends on range of max-min), then visualisation threshold is set also mear to min value a but a bit higher as a "invalid data", then setup scale and result you can see above. 

In circle are also sometimes invalid data, so then are displayed transparent spots here.

Then i need to determine x/y from chart, and you answer was really helpfull for me, because for now i did not have solution. 
Anyway, i am really thanks for your perfetcs inputs and for scottplot, its an amazing tool.

Now i see, you don't want using NaN values in data, and how about nullble double? Is possible, that plot will just ignorate nullable doubles? Perhaps you already coming through this discussion, i am just curious :)

Because i noticed now in version 1.0.0.47 runchart (plot) does not accept NaN (it was accepted in 42 version, which i used before) and i think it will be great to have possibility in chart for null data (just ignore record and create a gap in plot) But maybe the realization is not so easy as my imagination :)

Thanks guys ;)
kali

**Comment by Benny121221** [comment]
> Now i see, you don't want using NaN values in data, and how about nullble double? Is possible, that plot will just ignorate nullable doubles? Perhaps you already coming through this discussion, i am just curious :)

This isn't _too_ difficult, in #457 I dismissed the idea, although I cannot find why. I think the transparency threshold idea is tailored more to overlaying a heatmap over an image, for example in this image from #424
![image](https://user-images.githubusercontent.com/8635304/102724690-8ced7300-42ce-11eb-9cc6-5ed670137c19.png)

The low intensity values are transparent to allow whats beneath show through. In your case, it's not that you want to show only high intensity values, it's that you don't have values there, and the transparency thresholding is a workaround. So, I decided to add it to #668.
